plugins/modules/dcnm_image_upgrade.py

```diff
@@ -0,0 +1,5313 @@
+        params_spec[section][sub_section]["config_reload"] ["default"] = False
+
+        params_spec[section][sub_section]["write_erase"] = {}
+        params_spec[section][sub_section]["write_erase"] ["required"] = False
+        params_spec[section][sub_section]["write_erase"] ["type"] = "bool"
+        params_spec[section][sub_section]["write_erase"] ["default"] = False
+
+        sub_section = "package"
+        params_spec[section][sub_section] = {}
+        params_spec[section][sub_section]["required"] = False
+        params_spec[section][sub_section]["type"] = "dict"
+        params_spec[section][sub_section]["default"] = {}
+
+        params_spec[section][sub_section]["install"] = {}
+        params_spec[section][sub_section]["install"] ["required"] = False
+        params_spec[section][sub_section]["install"] ["type"] = "bool"
+        params_spec[section][sub_section]["install"] ["default"] = False
+
+        params_spec[section][sub_section]["uninstall"] = {}
+        params_spec[section][sub_section]["uninstall"] ["required"] = False
+        params_spec[section][sub_section]["uninstall"] ["type"] = "bool"
+        params_spec[section][sub_section]["uninstall"] ["default"] = False
+
+        return copy.deepcopy(params_spec)
+
+    def validate_input(self):
+        """
+        Caller: main()
+
+        Validate the playbook parameters
+        """
+        state = self.params["state"]
+        self.log_msg(f"{self.class_name}.validate_input: state: {state}")
+
+        if state not in ["merged", "deleted", "query"]:
+            msg = f"This module supports deleted, merged, and query states. Got state {state}"
+            self.module.fail_json(msg)
+
+        if state == "merged":
+            self.log_msg(f"{self.class_name}.validate_input: call _validate_input_for_merged_state()")
+            self._validate_input_for_merged_state()
+            return
+        if state == "deleted":
+            self._validate_input_for_deleted_state()
+            return
+        if state == "query":
+            self._validate_input_for_query_state()
+            return
+
+    def _validate_input_for_merged_state(self):
+        """
+        Caller: self.validate_input()
+
+        Validate that self.config contains appropriate values for merged state
+        """
+        if not self.config:
+            msg = "config: element is mandatory for state merged"
+            self.module.fail_json(msg)
+
+        params_spec = self._build_params_spec_for_merged_state()
+
+        self.log_msg(f"{self.class_name}._validate_input_for_merged_state: params_spec: {params_spec}")
+        self.log_msg(f"{self.class_name}._validate_input_for_merged_state: self.config: {self.config}")
+        valid_params, invalid_params = validate_list_of_dicts(
+            self.config.get("switches"), params_spec, self.module
+        )
+        # We're not using self.validated. Keeping this to avoid
```

Comment:
Hi Mallik, I did look into validate_list_of_dicts() but it's been a while since then.  I know there was a reason I didn't use valid_params (I also looked into the other standard Ansible validator methods documented on the ansible dev site).  Please give me a few days to review my earlier reasoning and I'll get back with you on this.